### PR TITLE
feat: added transitive dependencies

### PR DESCRIPTION
**Pull request by Obliviax**
from pablo-snyk:

In order to extend our vulnerability scanning feature to support child dependencies for [npm packages](https://docs.npmjs.com/cli/v6/configuring-npm/package-json) as described in [snyk issue 5, omitted] , this pull request updates the package service to return all nested dependencies on the internal /package endpoint for consumption by vulnerability service instead of just returning the versions for the first level.

It supports a GET request to the /package/:name/:version endpoint and will return a JSON structure representing the full tree of dependencies. We will always return the latest matching version of a package supported to mimic the behavior of a fresh npm install.

Testing

It can be tested locally by making a request for a package with sub-dependencies e.g. react@16.13.0

curl -s http://localhost:3000/package/react/16.13.0 | jq .
Related Ticket
[omitted]

### CODE REVIEW

File: packages/modules/npm.py [hunk 1]
```diff
@@ -1,21 +1,44 @@
...
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
+    url = f"{NPM_REGISTRY_URL}/{name}"
 
-    response = requests.get(url)
-    npm_package = response.json()
+    npm_package = requests.get(url).json()
+    versions = list(npm_package["versions"].keys())
+    version = semver.min_satisfying(versions, range)
```
Comment (Obliviax) [inline packages/modules/npm.py:14]: Raise an error if version is None; e.g., no matching result. Alternatively, update to return an empty dependency and continue the analysis. An error seems preferable because that would fail in production.

File: packages/modules/npm.py [hunk 2]
```diff
@@ -1,21 +1,44 @@
...
 
 
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
+    url = f"{NPM_REGISTRY_URL}/{name}"
 
-    response = requests.get(url)
-    npm_package = response.json()
+    npm_package = requests.get(url).json()
```
Comment (Obliviax) [inline packages/modules/npm.py:12]: before requesting versions, check if ver_range begins with "npm:".  This indicates an alias, in which case want dependencies for the alias target name and version, but to store and cache them with the alias's name and ver_range.

File: packages/modules/npm.py [hunk 3]
```diff
@@ -1,21 +1,44 @@
...
 
 from packages.models import VersionedPackage
 
 NPM_REGISTRY_URL = "https://registry.npmjs.org"
 
 
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
+    url = f"{NPM_REGISTRY_URL}/{name}"
```
Comment (Obliviax) [inline packages/modules/npm.py:11]: Provide input sanitation.  For example, no empty package names, and only permitted characters.

File: packages/modules/npm.py (hunk 2 above)
Comment (Obliviax) [inline packages/modules/npm.py:12]: requests.get() should specify a connection timeout so it doesn't run forever.  timeout=(5,10) is standard and gives sufficient API fetch time.

Comment (Obliviax) [inline packages/modules/npm.py:12]: 1. Put a try/except around this to handle ConnectionError.  We could have a retry system but the simple approach is report network issues back to the user and let them rerun later.
2. Handle non 20x status code (not npm_response.ok).

File: packages/modules/npm.py [hunk 4]
```diff
@@ -1,21 +1,44 @@
...
+    version_record = npm_package["versions"][version]
 
     package = VersionedPackage(
-        name=npm_package["name"],
-        description=npm_package["description"],
-        version=npm_package["version"],
-        dependencies=npm_package["dependencies"],
+        name=version_record["name"],
+        version=version_record["version"],
+        description=version_record["description"],
```
Comment (Obliviax) [inline packages/modules/npm.py:20]: description may not exist and would crash.  
Suggesting: description=version_record.get("description", ""),

File: packages/modules/npm.py [hunk 5]
```diff
@@ -1,21 +1,44 @@
...
+    dependencies = version_record.get("dependencies", {})
+
+    package.dependencies = [
+        get_package(name=dep_name, range=dep_range) for dep_name, dep_range in dependencies.items()
+    ]
 
     return package
+
+
+def request_package(name: str, range: str) -> tuple[VersionedPackage, dict]:
```
Comment (Obliviax) [inline packages/modules/npm.py:31]: Remove if unused.

File: packages/urls.py [hunk 6]
```diff
@@ -4,5 +4,5 @@
 
 urlpatterns = [
     path("<str:package_name>", views.PackageView.as_view()),
-    path("<str:package_name>/<str:version>", views.PackageView.as_view()),
```
Comment (Obliviax) [inline packages/urls.py:7]: API paths should be versioned, which helps forward and backward compatibility.  /api/v1/package/<pkg> if we can overwrite the old version; however, if it's in production consider keeping it and making this /api/v1/package/<pkg>.

File: packages/urls.py [hunk 7]
```diff
@@ -4,5 +4,5 @@
...
+    path("<str:package_name>/<str:range>", views.PackageView.as_view()),
```
Comment (Obliviax) [inline packages/urls.py:7]: How does this handle namespaces in the package name, e.g., <at>yarnpkg/core@^2.4.0 ?  In my test the additional / broke the path.  Perhaps a regex approach could handle the escaped @, /, and ^ characters while keeping the <name>/<version> format.

**Review by Obliviax — COMMENTED** [review]
Addresses the main point of recursing on dependencies to return a nested JSON of them.  Needs more handling and tests around edge cases and performance. Please see comments for minor improvements.

A multi-threaded approach should be considered. One thought is a BFS traversal with each thread loading a dependency since siblings are independent. They can have a shared cache with appropriate locking.
